src/Compilers/Core/MSBuildTask/LocateEditorConfigFiles.cs

```diff
@@ -0,0 +1,161 @@
+
+                        // We will attempt to determine if this is a root .editorconfig, so we can stop enumerating farther if necessary.
+                        // We will do no further interpretation of the .editorconfig.
+
+                        isRootEditorConfig = FileIsRootEditorConfig(editorConfigFilePath);
+                    }
+                }
+                catch (IOException exception)
+                {
+                    // We'll log the error and continue; the "it failed" is handled in Execute().
+                    Log.LogErrorFromResources(nameof(ErrorString.General_UnableToReadFile), editorConfigFilePath, exception.Message);
+                }
+
+                if (!isRootEditorConfig && directory.Parent != null)
+                {
+                    addNewDirectoryIfNotAlreadyAdded(directory.Parent);
+                }
+            }
+
+            EditorConfigFiles = editorConfigFiles.ToArray();
+        }
+
+        // These Regexes are the regexes from Microsoft.VisualStudio.CodingConventions.EditorConfig.
+        private static Regex s_sectionMatcher = new Regex(@"^\s*\[(([^#;]|\\#|\\;)+)\]\s*([#;].*)?$", RegexOptions.Compiled);
+        private static Regex s_propertyMatcher = new Regex(@"^\s*([\w\.\-_]+)\s*[=:]\s*(.*?)\s*([#;].*)?$", RegexOptions.Compiled);
+
+        internal static bool FileIsRootEditorConfig(string editorConfigFilePath)
+        {
+            using (FileStream fileStream = OpenFile(editorConfigFilePath))
+            using (StreamReader reader = new StreamReader(fileStream))
+            {
+                string line;
+                while ((line = reader.ReadLine()) != null)
+                {
+                    // Once we're in a section, we don't need to go looking for root = true anymore
+                    if (s_sectionMatcher.IsMatch(line))
+                    {
+                        return false;
+                    }
+
+                    var propMatches = s_propertyMatcher.Matches(line);
+                    if (propMatches.Count == 1 && propMatches[0].Groups.Count > 2)
+                    {
+                        var key = propMatches[0].Groups[1].Value;
+                        var value = propMatches[0].Groups[2].Value;
+
+                        if (key == "root")
+                        {
+                            return value == "true";
+                        }
+                    }
+                }
+            }
+
+            return false;
+        }
+
+        private static FileStream OpenFile(string editorConfigFilePath)
+        {
+            try
+            {
+                return new FileStream(editorConfigFilePath, FileMode.Open, FileAccess.Read, FileShare.Read);
+            }
+            catch (IOException)
+            {
+                throw;
+            }
+            catch (Exception e)
+            {
+                throw new IOException(e.Message, e);
```

Comment:
Catch execption so we can throw IOException so we can catch IOException.  So much complex.  Why not justn catch Exception?